pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -981,6 +890,96 @@ func (c *Controller) updateEDS(ep *v1.Endpoints, event model.Event) {
 	_ = c.XDSUpdater.EDSUpdate(c.ClusterID, string(hostname), ep.Namespace, endpoints)
 }
 
+func (c *Controller) updateEDSSlice(slice *discoveryv1alpha1.EndpointSlice, event model.Event) {
+	svcName := slice.Labels[discoveryv1alpha1.LabelServiceName]
+	hostname := kube.ServiceHostname(svcName, slice.Namespace, c.domainSuffix)
+	mixerEnabled := c.Env != nil && c.Env.Mesh != nil && (c.Env.Mesh.MixerCheckServer != "" || c.Env.Mesh.MixerReportServer != "")
+
+	endpoints := make([]*model.IstioEndpoint, 0)
+	if event != model.EventDelete {
+		for _, e := range slice.Endpoints {
+			if e.Conditions.Ready != nil && !*e.Conditions.Ready {
+				// Ignore not ready endpoints
+				continue
+			}
+			for _, a := range e.Addresses {
+				pod := c.pods.getPodByIP(a)
+				if pod == nil {
+					// This can not happen in usual case
+					if e.TargetRef != nil && e.TargetRef.Kind == "Pod" {
+						log.Warnf("Endpoint without pod %s %s.%s", a, svcName, slice.Namespace)
+
+						if c.Env != nil {
+							c.Env.PushContext.Add(model.EndpointNoPod, string(hostname), nil, a)
+						}
+						// TODO: keep them in a list, and check when pod events happen !
+						continue
+					}
+					// For service without selector, maybe there are no related pods
+				}
+
+				var labels map[string]string
+				locality, sa, uid := "", "", ""
+				if pod != nil {
+					locality = c.GetPodLocality(pod)
```

Comment:
We donot  need this, just get from the endpointslice.

Reply:
Good point